### ISSUE TITLE
[miniLLM] About KL divergence

### ISSUE THREAD

**Issue by wutaiqiang** [microsoft/LMOps#116]
hi, thanks for your nice paper.

For KL, the code is 
> def get_rev_kl(log_p, log_q, mask):
    log_ratio = (log_p - log_q) * mask
    kl = log_ratio.float().exp() - 1 - log_ratio
    return kl

may I know why log_ratio.float().exp() - 1 - log_ratio ? Thanks

**Comment by wutaiqiang** [comment]
@t1101675

**Comment by t1101675** [comment]
It is a low-variance estimation of KL divergence http://joschu.net/blog/kl-approx.html ,

**Comment by wutaiqiang** [comment]
Thanks for your response. Why not just use F.kl_div function？

**Comment by wutaiqiang** [comment]
By the way, for KD baseline, you code is:

>         teacher_probs = F.softmax(teacher_logits, dim=-1, dtype=torch.float32)
        inf_mask = torch.isinf(logits)
        logprobs = F.log_softmax(logits, dim=-1, dtype=torch.float32)
        prod_probs = torch.masked_fill(teacher_probs * logprobs, inf_mask, 0)
        x = torch.sum(prod_probs, dim=-1).view(-1)
        mask = (no_model_batch["label"] != -100).int()
        distil_loss = -torch.sum(x * mask.view(-1), dim=0) / torch.sum(mask.view(-1), dim=0)

seems that you only calculate sum(plog (q)), yes p*logp is a constant and can be omitted, but usually, we would use F.kl_div rather.

**Comment by t1101675** [comment]
We implement KLD by ourselves for the convenience of debugging and adapting to tensor parallel cases. Of course, it can be replaced with F.kd_div.

**Comment by wutaiqiang** [comment]
Thanks for your kind reply
